vec/src/Data/Vec/Lazy/Inline.hs

```diff
@@ -520,6 +526,44 @@ ifoldr = getIFoldr $ N.induction1 start step where
 
 newtype IFoldr a n b = IFoldr { getIFoldr :: (Fin n -> a -> b -> b) -> b -> Vec n a -> b }
 
+scanr :: forall a b n. N.SNatI n => (a -> b -> b) -> b -> Vec n a -> Vec ('S n) b
+scanr f z = getScanr $ N.induction1 start step where
+    start :: Scanr a 'Z b
+    start = Scanr $ \_ -> singleton z
+
+    step :: Scanr a m b -> Scanr a ('S m) b
+    step (Scanr go) = Scanr $ \(x ::: xs) -> case go xs of
+        ys@(y ::: _) -> f x y ::: ys
+
+newtype Scanr a n b = Scanr { getScanr :: Vec n a -> Vec ('S n) b }
+
+scanl :: forall a b n. N.SNatI n => (b -> a -> b) -> b -> Vec n a -> Vec ('S n) b
+scanl f = getScanl $ N.induction1 start step where
+    start :: Scanl a 'Z b
+    start = Scanl $ \z VNil -> singleton z
+
+    step :: Scanl a m b -> Scanl a ('S m) b
+    step (Scanl go) = Scanl $ \(!acc) (x ::: xs) -> acc ::: go (f acc x) xs
+
+newtype Scanl a n b = Scanl { getScanl :: b -> Vec n a -> Vec ('S n) b }
+
+scanr1 :: forall a n. N.SNatI n => (a -> a -> a) -> Vec n a -> Vec n a
+scanr1 f = getScanr1 $ N.induction1 start step where
+    start :: Scanr1 'Z a
+    start = Scanr1 $ \_ -> VNil
+
+    step :: forall m. N.SNatI m => Scanr1 m a -> Scanr1 ('S m) a
+    step (Scanr1 go) = Scanr1 $ \(x ::: xs) -> case N.snat :: N.SNat m of
```

Comment:
This doesn't feel right. You shouldn't need to check length in the step case. I can take a look myself if you cannot find a way to avoid it.

Reply:
In `scanr1`, the last element is special (it is the zero element), and so the `0 -> 1` step and the `m -> m + 1` step (where `m > 0`) are different. But yeah please let me know if there's a better way to write this.